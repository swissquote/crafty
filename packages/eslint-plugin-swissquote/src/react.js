const { addMissingRules } = require("./utils");

// EcmaScript 6 specific configuration
module.exports = {
  settings: {
    "import/resolver": {
      node: {
        extensions: [".js", ".jsx", ".json"]
      }
    },
    react: {
      version: "detect"
    }
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true
    }
  },
  rules: {
    // Recommended rules override
    "@swissquote/swissquote/react/display-name": 0, // Disabled as it generates false positives
    "@swissquote/swissquote/react/jsx-no-duplicate-props": [
      "error",
      { ignoreCase: true }
    ],
    "@swissquote/swissquote/react/no-deprecated": "warn",
    "@swissquote/swissquote/react/prop-types": [
      "warn",
      { skipUndeclared: true }
    ],

    // Swissquote Rules
    "@swissquote/swissquote/react/jsx-handler-names": "error",
    "@swissquote/swissquote/react/jsx-pascal-case": "error",
    "@swissquote/swissquote/react/no-did-mount-set-state": "error",
    "@swissquote/swissquote/react/no-did-update-set-state": "error",
    "@swissquote/swissquote/react/no-redundant-should-component-update":
      "error",
    "@swissquote/swissquote/react/no-typos": "error",
    "@swissquote/swissquote/react/no-unused-state": "error",
    "@swissquote/swissquote/react/no-will-update-set-state": "error",
    "@swissquote/swissquote/react/prefer-es6-class": ["error", "always"],
    "@swissquote/swissquote/react/prefer-stateless-function": [
      "error",
      { ignorePureComponents: true }
    ],
    "@swissquote/swissquote/react/void-dom-elements-no-children": "error"
  }
};

// Add all recommended configurations from eslint-plugin-react
addMissingRules(
  require("eslint-plugin-react").configs.recommended.rules,
  module.exports.rules
);

// Disable all the rules from eslint-plugin-react that are handled by Prettier
addMissingRules(
  require("eslint-config-prettier/react").rules,
  module.exports.rules
);
