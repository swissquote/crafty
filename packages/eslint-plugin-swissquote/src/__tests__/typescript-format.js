/* global describe, it, expect, jest */

const { prepareESLint, lint } = require("../../test_utils");

const engine = prepareESLint("format");

it("Fails on badly formatted TypeScript code", async () => {
  const result = await lint(
    engine,
    `
module.exports = function initJS  (gulp, config: {}, watchers): string[] {
  const js = config.js,
     jsTasks: string[] = [];

  for (const name in js) {
    if (!js.hasOwnProperty(name)) {
      continue;
      }

    const taskName = \`js_\${name}\`;

    if (!compileWithWebpack(js[name])) {
      gulp.task(taskName, jsTaskES5(gulp, config, watchers, js[name]));
        watchers.add(js[name].watch || js[name].source, taskName);
    }

    jsTasks.push( taskName );
  }

  gulp.task("js", jsTasks);

  return ["js"];
};
`,
    "file.ts"
  );

  expect(result.messages).toMatchSnapshot();
  expect(result.warningCount).toBe(0);
  expect(result.errorCount).toBe(5);
});

it("Works with complex types", async () => {
  const result = await lint(
    engine,
    `
import * as React from "react";

type Writeable<T> = { -readonly [P in keyof T]-?: T[P] };

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

function assignRef<T>(ref: T, setter: React.Ref<T>) {
  if (!setter || !ref) {
    return;
  }

  if (typeof setter === "function") {
    setter(ref);
    return;
  }

  if (setter.hasOwnProperty("current")) {
    (setter as Writeable<React.RefObject<T>>).current = ref;
    return;
  }

  throw new Error(
    "Your ref must be a function or an object created with React.createRef"
  );
}

export type triggerFunction = (options: {
  isOpen: boolean;
  caret: React.ReactNode;
  props: object;
}) => React.ReactNode;

export interface ButtonGroupProps extends React.HTMLAttributes<HTMLDivElement> {
  innerRef?: React.Ref<HTMLDivElement>;
}

export interface SplitButtonProps extends Omit<ButtonGroupProps, "innerRef"> {
  other: React.ReactNode;
  open?: boolean;
  trigger: React.ReactNode | triggerFunction;

  /**
   * Ref to get the component
   */
  innerRef?: React.Ref<HTMLElement>;
}

export interface SplitButtonState {
  open?: boolean;
}

export default class SplitButton extends React.Component<
  SplitButtonProps,
  SplitButtonState
> {
  constructor(props) {
    super(props);

    this.state = {
      open: props.open || false
    };
  }

  private target: Element;

  handleRef = node => {
    this.target = getElement(node);
    assignRef(this.target, this.props.innerRef);
  };

  render(trigger) {
    const { trigger } = this.props;
    if (typeof trigger !== "function") {
      return <button innerRef={this.handleRef}>{trigger}</button>;
    }

    return (trigger as triggerFunction)({
      props: { innerRef: this.handleRef },
      isOpen: this.state.open,
      caret: <span className={this.props.classes("Caret")} />
    });
  }
}
`,
    "Component.tsx"
  );

  expect(result.warningCount).toBe(0);
  expect(result.errorCount).toBe(0);
});
