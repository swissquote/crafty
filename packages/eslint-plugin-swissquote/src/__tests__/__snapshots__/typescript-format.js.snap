// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fails on badly formatted TypeScript code 1`] = `
Array [
  Object {
    "column": 33,
    "endColumn": 35,
    "endLine": 1,
    "fix": Object {
      "range": Array [
        32,
        34,
      ],
      "text": "",
    },
    "line": 1,
    "message": "Delete \`··\`",
    "nodeType": null,
    "ruleId": "@swissquote/swissquote/prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 1,
    "endColumn": 2,
    "endLine": 3,
    "fix": Object {
      "range": Array [
        99,
        100,
      ],
      "text": "",
    },
    "line": 3,
    "message": "Delete \`·\`",
    "nodeType": null,
    "ruleId": "@swissquote/swissquote/prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 1,
    "endColumn": 3,
    "endLine": 8,
    "fix": Object {
      "range": Array [
        208,
        210,
      ],
      "text": "",
    },
    "line": 8,
    "message": "Delete \`··\`",
    "nodeType": null,
    "ruleId": "@swissquote/swissquote/prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 7,
    "endColumn": 9,
    "endLine": 14,
    "fix": Object {
      "range": Array [
        372,
        374,
      ],
      "text": "",
    },
    "line": 14,
    "message": "Delete \`··\`",
    "nodeType": null,
    "ruleId": "@swissquote/swissquote/prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 18,
    "endColumn": 28,
    "endLine": 17,
    "fix": Object {
      "range": Array [
        457,
        467,
      ],
      "text": "taskName",
    },
    "line": 17,
    "message": "Replace \`·taskName·\` with \`taskName\`",
    "nodeType": null,
    "ruleId": "@swissquote/swissquote/prettier/prettier",
    "severity": 2,
  },
]
`;
